File: app/frontend/lib/employees.spec.js
import { afterEach, describe, expect, it, vi } from "vitest"
import {
  addCompensation,
  compensationChangeErrors,
  compensationErrors,
  directoryFilterErrors,
  employeeIdentityErrors,
  fieldErrorText,
  firstApiFieldError,
  getEmployee,
  importErrors,
  importToastTitle,
  listEmployees,
  importEmployees,
  offboardEmployee,
  offboardErrors,
  onboardEmployee,
  onboardErrors,
  updateCompensation,
  updateEmployee,
  deleteCompensation,
  destroyEmployee,
  exportEmployees,
  rehireEmployee
} from "./employees.js"

describe("directoryFilterErrors", () => {
  it("accepts blank filters", () => {
    expect(directoryFilterErrors({})).toEqual({})
  })

  it("rejects an unknown type, status, or country", () => {
    expect(directoryFilterErrors({ country: "USA", type: "contractor-plus", status: "onboarding" })).toEqual({
      country: "unknown country",
      type: "unknown type",
      status: "unknown status"
    })
  })

  it("rejects a search that is too long", () => {
    expect(directoryFilterErrors({ q: "a".repeat(256) })).toEqual({ q: "q is too long" })
  })

  it("accepts a comma-separated country list", () => {
    expect(directoryFilterErrors({ country: "GB,US" })).toEqual({})
  })
})

describe("onboardErrors", () => {
  const valid = {
    first_name: "Ada",
    last_name: "Lovelace",
    email: "ada@acme.test",
    country: "GB",
    department: "engineering",
    employment_type: "full-time",
    started_on: "2024-01-01",
    compensation: { base_amount: 80000, currency: "GBP", pay_period: "annual", effective_date: "2024-01-01" }
  }

  it("accepts a complete hire", () => {
    expect(onboardErrors(valid)).toEqual({})
  })

  it("requires compensation and identity fields", () => {
    expect(onboardErrors({})).toMatchObject({
      first_name: "Enter a first name",
      email: "Enter an email",
      compensation: "compensation is required"
    })
  })
})

describe("compensationErrors", () => {
  it("requires hours for hourly pay", () => {
    expect(compensationErrors({ base_amount: 40, currency: "USD", pay_period: "hourly" })).toEqual({
      hours_per_week: "hours_per_week is required for hourly pay"
    })
  })

  it("rejects an unknown currency or pay period", () => {
    expect(compensationErrors({ base_amount: 1, currency: "JPY", pay_period: "weekly" })).toEqual({
      currency: "unknown currency",
      pay_period: "unknown pay period"
    })
  })

  it("rejects daily and monthly pay on new or edited rows", () => {
    expect(compensationErrors({ base_amount: 1, currency: "USD", pay_period: "monthly" })).toEqual({
      pay_period: "unknown pay period"
    })
    expect(compensationErrors({ base_amount: 1, currency: "USD", pay_period: "daily" })).toEqual({
      pay_period: "unknown pay period"
    })
  })
})

describe("compensationChangeErrors", () => {
  it("requires an effective date", () => {
    expect(compensationChangeErrors({ base_amount: 1, currency: "USD", pay_period: "annual" })).toEqual({
      effective_date: "Enter an effective date"
    })
  })
})

describe("fieldErrorText", () => {
  it("reads the first API detail message", () => {
    expect(fieldErrorText([ "has already been taken" ])).toBe("has already been taken")
    expect(firstApiFieldError({ email: [ "has already been taken" ] })).toBe("has already been taken")
  })
})

describe("employeeIdentityErrors", () => {
  const valid = {
    first_name: "Ada",
    last_name: "Lovelace",
    email: "ada@acme.test",
    country: "GB",
    department: "engineering",
    employment_type: "full-time",
    started_on: "2024-01-01"
  }

  it("accepts identity fields without compensation", () => {
    expect(employeeIdentityErrors(valid)).toEqual({})
  })

  it("requires name, email, and start date", () => {
    expect(employeeIdentityErrors({})).toMatchObject({
      first_name: "Enter a first name",
      email: "Enter an email",
      started_on: "Enter a start date"
    })
  })
})

describe("offboardErrors", () => {
  it("requires an ISO leave date", () => {
    expect(offboardErrors({})).toEqual({ left_on: "left_on is required" })
    expect(offboardErrors({ left_on: "June 1" })).toEqual({ left_on: "left_on is invalid" })
  })

  it("rejects a leave date before started_on", () => {
    expect(offboardErrors({ left_on: "2023-12-01", started_on: "2024-01-01" })).toEqual({
      left_on: "must be on or after started_on"
    })
  })
})

describe("importErrors", () => {
  it("requires a CSV under the size cap", () => {
    expect(importErrors()).toEqual({ file: "Choose a CSV file" })
    expect(importErrors(new File([ "x" ], "people.txt", { type: "text/plain" }))).toEqual({
      file: "upload a CSV file"
    })
    expect(importErrors(new File([ "a" ], "people.csv", { type: "text/csv" }))).toEqual({})
  })

  it("rejects a file over the size cap", () => {
    expect(importErrors({ name: "people.csv", type: "text/csv", size: 5 * 1024 * 1024 + 1 })).toEqual({
      file: "file is too large"
    })
  })
})

describe("importToastTitle", () => {
  it("names imported and updated counts", () => {
    expect(importToastTitle({ data: { employees: 2 } })).toBe("Imported 2 employees")
    expect(importToastTitle({ employees: 0, updated: 2 })).toBe("Imported 0 employees · updated 2")
    expect(importToastTitle({ employees: 1 })).toBe("Imported 1 employee")
  })
})

describe("employee requests", () => {
  afterEach(() => {
    vi.unstubAllGlobals()
    vi.restoreAllMocks()
  })

  it("lists employees with query params", async () => {
    const fetchMock = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ employees: [], pagination: { page: 1 } }), { status: 200 })
    )
    vi.stubGlobal("fetch", fetchMock)

    await listEmployees({ country: "GB", q: "ada", page: 2 })

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees?country=GB&q=ada&page=2",
      expect.objectContaining({ credentials: "same-origin" })
    )
  })

  it("does not call the API when directory filters are invalid", async () => {
    const fetchMock = vi.fn()
    vi.stubGlobal("fetch", fetchMock)

    await expect(listEmployees({ type: "contractor-plus" })).rejects.toThrow("unknown type")
    expect(fetchMock).not.toHaveBeenCalled()
  })

  it("does not call the API when q is too long", async () => {
    const fetchMock = vi.fn()
    vi.stubGlobal("fetch", fetchMock)

    await expect(listEmployees({ q: "a".repeat(256) })).rejects.toThrow("q is too long")
    expect(fetchMock).not.toHaveBeenCalled()
  })

  it("loads an employee profile", async () => {
    const fetchMock = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ data: { employee: { id: "emp-1" } } }), { status: 200 })
    )
    vi.stubGlobal("fetch", fetchMock)

    await getEmployee("emp-1")

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees/emp-1",
      expect.objectContaining({ credentials: "same-origin" })
    )
  })

  it("onboards an employee", async () => {
    const fetchMock = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ employee: { email: "ada@acme.test" } }), { status: 201 })
    )
    vi.stubGlobal("fetch", fetchMock)
    const payload = {
      first_name: "Ada",
      last_name: "Lovelace",
      email: "ada@acme.test",
      country: "GB",
      department: "engineering",
      employment_type: "full-time",
      started_on: "2024-01-01",
      compensation: { base_amount: 80000, currency: "GBP", pay_period: "annual" }
    }

    await onboardEmployee(payload)

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees",
      expect.objectContaining({ method: "POST", body: JSON.stringify(payload) })
    )
  })

  it("does not call the API when a pay change is missing an effective date", async () => {
    const fetchMock = vi.fn()
    vi.stubGlobal("fetch", fetchMock)

    await expect(addCompensation("emp-1", { base_amount: 1, currency: "USD", pay_period: "annual" })).rejects.toThrow(
      "Enter an effective date"
    )
    expect(fetchMock).not.toHaveBeenCalled()
  })

  it("attaches API field details to the thrown error", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn().mockResolvedValue(
        new Response(
          JSON.stringify({
            error: {
              code: "validation_failed",
              message: "validation failed",
              details: { email: [ "has already been taken" ] }
            }
          }),
          { status: 422 }
        )
      )
    )

    await expect(
      onboardEmployee({
        first_name: "Ada",
        last_name: "Lovelace",
        email: "ada@acme.test",
        country: "GB",
        department: "engineering",
        employment_type: "full-time",
        started_on: "2024-01-01",
        compensation: { base_amount: 80000, currency: "GBP", pay_period: "annual" }
      })
    ).rejects.toMatchObject({
      message: "validation failed",
      details: { email: [ "has already been taken" ] }
    })
  })

  it("updates an employee", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response("{}", { status: 200 }))
    vi.stubGlobal("fetch", fetchMock)
    const payload = {
      first_name: "Grace",
      last_name: "Hopper",
      email: "grace@acme.test",
      country: "US",
      department: "sales",
      employment_type: "full-time",
      started_on: "2024-01-01"
    }

    await updateEmployee("emp-1", payload)

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees/emp-1",
      expect.objectContaining({ method: "PATCH", body: JSON.stringify(payload) })
    )
  })

  it("does not call the API when an update is missing a name", async () => {
    const fetchMock = vi.fn()
    vi.stubGlobal("fetch", fetchMock)

    await expect(updateEmployee("emp-1", { last_name: "Hopper" })).rejects.toThrow("Enter a first name")
    expect(fetchMock).not.toHaveBeenCalled()
  })

  it("records a compensation change", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response("{}", { status: 201 }))
    vi.stubGlobal("fetch", fetchMock)
    const payload = { base_amount: 90000, currency: "GBP", pay_period: "annual", effective_date: "2025-04-01" }

    await addCompensation("emp-1", payload)

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees/emp-1/compensation_records",
      expect.objectContaining({ method: "POST", body: JSON.stringify(payload) })
    )
  })

  it("imports a CSV without forcing a JSON content type", async () => {
    const fetchMock = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ data: { employees: 1 } }), { status: 200 })
    )
    vi.stubGlobal("fetch", fetchMock)
    const file = new File([ "first_name\nAda" ], "people.csv", { type: "text/csv" })

    await importEmployees(file)

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees/import",
      expect.objectContaining({ method: "POST", body: expect.any(FormData) })
    )
    expect(fetchMock.mock.calls[0][1].headers["Content-Type"]).toBeUndefined()
  })

  it("does not call the API when no file is chosen", async () => {
    const fetchMock = vi.fn()
    vi.stubGlobal("fetch", fetchMock)

    await expect(importEmployees()).rejects.toThrow("Choose a CSV file")
    expect(fetchMock).not.toHaveBeenCalled()
  })

  it("corrects a pay row", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response("{}", { status: 200 }))
    vi.stubGlobal("fetch", fetchMock)
    const payload = { base_amount: 81000, currency: "GBP", pay_period: "annual", effective_date: "2024-01-01" }

    await updateCompensation("emp-1", "comp-1", payload)

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees/emp-1/compensation_records/comp-1",
      expect.objectContaining({ method: "PATCH", body: JSON.stringify(payload) })
    )
  })

  it("deletes a pay row", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response("{}", { status: 200 }))
    vi.stubGlobal("fetch", fetchMock)

    await deleteCompensation("emp-1", "comp-1")

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees/emp-1/compensation_records/comp-1",
      expect.objectContaining({ method: "DELETE" })
    )
  })

  it("rehires and deletes an employee", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response("{}", { status: 200 }))
    vi.stubGlobal("fetch", fetchMock)

    await rehireEmployee("emp-1")
    await destroyEmployee("emp-1")

    expect(fetchMock).toHaveBeenNthCalledWith(
      1,
      "/api/v1/employees/emp-1/rehire",
      expect.objectContaining({ method: "PATCH" })
    )
    expect(fetchMock).toHaveBeenNthCalledWith(
      2,
      "/api/v1/employees/emp-1",
      expect.objectContaining({ method: "DELETE" })
    )
  })

  it("exports the filtered directory", async () => {
    const fetchMock = vi.fn().mockResolvedValue(
      new Response("Name,Email\nAda,ada@acme.test", { status: 200, headers: { "Content-Type": "text/csv" } })
    )
    vi.stubGlobal("fetch", fetchMock)
    vi.stubGlobal("URL", { createObjectURL: () => "blob:export", revokeObjectURL: vi.fn() })

    await exportEmployees({ country: "GB" })

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees/export?country=GB",
      expect.objectContaining({
        headers: expect.objectContaining({ Accept: "text/csv" })
      })
    )
  })

  it("offboards an employee", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response("{}", { status: 200 }))
    vi.stubGlobal("fetch", fetchMock)

    await offboardEmployee("emp-1", { left_on: "2025-06-01" })

    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/employees/emp-1/offboard",
      expect.objectContaining({ method: "PATCH", body: JSON.stringify({ left_on: "2025-06-01" }) })
    )
  })

  it("raises an internal error message", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn().mockResolvedValue(
        new Response(JSON.stringify({ error: { code: "internal_error", message: "internal error" } }), { status: 500 })
      )
    )

    await expect(listEmployees()).rejects.toThrow("internal error")
  })

  it("raises the API error payload", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn().mockResolvedValue(
        new Response(JSON.stringify({ error: { code: "invalid_request", message: "already left" } }), { status: 422 })
      )
    )

    await expect(offboardEmployee("emp-1", { left_on: "2025-06-01" })).rejects.toThrow("already left")
  })
})

File: spec/requests/swagger/employees_spec.rb
require "swagger_helper"

RSpec.describe "Employees", type: :request do
  def onboard_body(**overrides)
    {
      first_name: "Ada",
      last_name: "Lovelace",
      email: "ada@acme.test",
      country: "GB",
      department: "engineering",
      employment_type: "full-time",
      started_on: "2024-01-01",
      compensation: {
        base_amount: 80_000,
        currency: "GBP",
        pay_period: "annual",
        effective_date: "2024-01-01",
        change_reason: "hire"
      }
    }.deep_merge(overrides)
  end

  path "/api/v1/employees" do
    get "List the directory" do
      tags "Employees"
      produces "application/json"
      description "Server-paginated. Filters use the directory index. `q` is trigram search on name and email."
      parameter name: :page, in: :query, required: false, schema: { type: :integer, minimum: 1 }
      parameter name: :per_page, in: :query, required: false, schema: { type: :integer, minimum: 1, maximum: 100 }
      parameter name: :limit, in: :query, required: false, schema: { type: :integer, minimum: 1, maximum: 100 }
      parameter name: :country, in: :query, required: false, schema: { type: :string, minLength: 2, maxLength: 2 }
      parameter name: :department, in: :query, required: false, schema: { type: :string }
      parameter name: :type, in: :query, required: false, schema: { type: :string, enum: Employee::EMPLOYMENT_TYPES }
      parameter name: :employment_type, in: :query, required: false, schema: { type: :string, enum: Employee::EMPLOYMENT_TYPES }
      parameter name: :status, in: :query, required: false, getter: :employment_status,
                schema: { type: :string, enum: Employee::STATUSES }
      parameter name: :q, in: :query, required: false, schema: { type: :string, maxLength: 255 }

      response "200", "page of employees" do
        schema "$ref" => "#/components/schemas/DirectoryPage"

        context "when listing the first page" do
          before do
            sign_in_hr
            create(:employee, first_name: "Ada", last_name: "Lovelace")
          end

          run_test! do |response|
            expect(api_meta.fetch("pagination")).to include("page" => 1, "limit" => 25)
          end
        end

        context "when paging past the last page" do
          let(:page) { 9 }
          before do
            sign_in_hr
            create(:employee)
          end

          run_test! do |response|
            expect(api_data.fetch("employees")).to eq([])
          end
        end

        context "when per_page is above the max" do
          let(:per_page) { 500 }
          before { sign_in_hr }

          run_test! do |response|
            expect(api_meta.dig("pagination", "limit")).to eq(100)
          end
        end

        context "when filtering and searching" do
          let(:country) { "gb" }
          let(:department) { "engineering" }
          let(:type) { "full-time" }
          let(:employment_status) { "active" }
          let(:q) { "ada" }
          before do
            sign_in_hr
            create(:employee, first_name: "Ada", country: "GB", department: "engineering")
            create(:employee, first_name: "Grace", country: "US", department: "sales", email: "grace@acme.test")
          end

          run_test! do |response|
            expect(api_data.fetch("employees").pluck("first_name")).to eq([ "Ada" ])
          end
        end

        context "when filtering by employment_type" do
          let(:employment_type) { "intern" }
          before do
            sign_in_hr
            create(:employee, employment_type: "intern")
            create(:employee, employment_type: "full-time", email: "ft@acme.test")
          end

          run_test! do |response|
            expect(api_data.fetch("employees").pluck("employment_type")).to eq([ "intern" ])
          end
        end
      end

      response "401", "not signed in" do
        schema "$ref" => "#/components/schemas/Error"
        run_test!
      end

      response "422", "invalid filter" do
        schema "$ref" => "#/components/schemas/Error"
        before { sign_in_hr }

        context "when type is unknown" do
          let(:type) { "contractor-plus" }

          run_test! do |response|
            expect(api_error).to include("code" => "invalid_request", "message" => "unknown type")
          end
        end

        context "when status is unknown" do
          let(:employment_status) { "onboarding" }

          run_test! do |response|
            expect(api_error).to include("message" => "unknown status")
          end
        end

        context "when country is not ISO-2" do
          let(:country) { "USA" }

          run_test! do |response|
            expect(api_error).to include("message" => "unknown country")
          end
        end

        context "when q is too long" do
          let(:q) { "a" * (DirectoryQuery::MAX_QUERY + 1) }

          run_test! do |response|
            expect(api_error).to include("message" => "q is too long")
          end
        end
      end

      response "500", "internal error" do
        schema "$ref" => "#/components/schemas/Error"
        before do
          sign_in_hr
          allow(DirectoryQuery).to receive(:new).and_raise(RuntimeError, "secret failure")
        end

        run_test! do |response|
          expect(api_error).to include("code" => "internal_error", "message" => "internal error")
        end
      end
    end

    post "Onboard an employee" do
      tags "Employees"
      consumes "application/json"
      produces "application/json"
      description "Creates an active employee and the first compensation row in one transaction."
      parameter name: :body, in: :body, schema: { "$ref" => "#/components/schemas/OnboardRequest" }

      response "201", "hired" do
        schema "$ref" => "#/components/schemas/HireResponse"
        let(:body) { onboard_body }
        before { sign_in_hr }

        run_test! do |response|
          expect(api_data.dig("employee", "status")).to eq("active")
          expect(api_data.dig("compensation_record", "change_reason")).to eq("hire")
        end
      end

      response "401", "not signed in" do
        schema "$ref" => "#/components/schemas/Error"
        let(:body) { onboard_body }
        run_test!
      end

      response "422", "invalid hire" do
        schema "$ref" => "#/components/schemas/ValidationError"
        before { sign_in_hr }

        context "when compensation is missing" do
          let(:body) { onboard_body.except(:compensation) }

          run_test! do |response|
            expect(api_error).to include("message" => "compensation is required")
          end
        end

        context "when compensation is empty" do
          let(:body) { onboard_body.merge(compensation: {}) }

          run_test! do |response|
            expect(api_error).to include("message" => "compensation is required")
          end
        end

        context "when the email is taken" do
          let(:body) { onboard_body }
          before { create(:employee, email: "ada@acme.test") }

          run_test! do |response|
            expect(api_error).to include("code" => "validation_failed", "message" => "validation failed")
          end
        end

        context "when first_name is blank" do
          let(:body) { onboard_body.merge(first_name: "") }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("first_name")
          end
        end

        context "when hourly pay has no hours" do
          let(:body) { onboard_body(compensation: { pay_period: "hourly", hours_per_week: nil }) }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("hours_per_week")
          end
        end

        context "when the currency is unsupported" do
          let(:body) { onboard_body(compensation: { currency: "JPY" }) }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("currency")
          end
        end

        context "when the country is not ISO-2" do
          let(:body) { onboard_body.merge(country: "USA") }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("country")
          end
        end
      end
    end
  end

  path "/api/v1/employees/{id}" do
    parameter name: :id, in: :path, schema: { type: :string, format: :uuid }

    get "Show an employee profile" do
      tags "Employees"
      produces "application/json"
      description "Current compensation plus newest-first effective-dated history. Amounts include annualised USD."

      response "200", "profile" do
        schema "$ref" => "#/components/schemas/EmployeeProfile"
        let(:employee) { create(:employee) }
        let(:id) { employee.id }
        before do
          sign_in_hr
          ExchangeRate.seed!(on: Date.new(2024, 1, 1))
          create(:compensation_record, employee: employee, change_reason: "hire")
        end

        run_test! do |response|
          expect(api_data.fetch("employee")).to include("id" => employee.id)
          expect(api_data.fetch("compensation_records")).not_to be_empty
        end
      end

      response "401", "not signed in" do
        schema "$ref" => "#/components/schemas/Error"
        let(:id) { create(:employee).id }
        run_test!
      end

      response "404", "unknown employee" do
        schema "$ref" => "#/components/schemas/Error"
        let(:id) { SecureRandom.uuid }
        before { sign_in_hr }

        run_test!
      end
    end

    patch "Update employee identity" do
      tags "Employees"
      consumes "application/json"
      produces "application/json"
      description "Edits profile fields. Status and leave date stay on offboard. Compensation stays on the pay-change endpoint."
      parameter name: :body, in: :body, schema: { "$ref" => "#/components/schemas/EmployeeUpdateRequest" }

      response "200", "updated profile" do
        schema "$ref" => "#/components/schemas/EmployeeProfile"
        let(:employee) { create(:employee, first_name: "Ada") }
        let(:id) { employee.id }
        let(:body) { { first_name: "Grace", department: "sales" } }
        before do
          sign_in_hr
          ExchangeRate.seed!(on: Date.new(2024, 1, 1))
          create(:compensation_record, employee: employee, change_reason: "hire")
        end

        run_test! do |response|
          expect(api_data.fetch("employee")).to include(
            "id" => employee.id, "first_name" => "Grace", "department" => "sales"
          )
        end
      end

      response "401", "not signed in" do
        schema "$ref" => "#/components/schemas/Error"
        let(:id) { create(:employee).id }
        let(:body) { { first_name: "Grace" } }
        run_test!
      end

      response "404", "unknown employee" do
        schema "$ref" => "#/components/schemas/Error"
        let(:id) { SecureRandom.uuid }
        let(:body) { { first_name: "Grace" } }
        before { sign_in_hr }

        run_test!
      end

      response "422", "cannot update" do
        schema "$ref" => "#/components/schemas/ValidationError"
        before { sign_in_hr }

        context "when no employee fields are sent" do
          let(:id) { create(:employee).id }
          let(:body) { {} }

          run_test! do |response|
            expect(api_error).to include("code" => "invalid_request", "message" => "no employee fields to update")
          end
        end

        context "when the email is taken" do
          let(:id) { create(:employee, email: "ada@acme.test").id }
          let(:body) { { email: "taken@acme.test" } }
          before { create(:employee, email: "taken@acme.test") }

          run_test! do |response|
            expect(api_error).to include("code" => "validation_failed")
          end
        end

        context "when first_name is blank" do
          let(:id) { create(:employee).id }
          let(:body) { { first_name: "" } }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("first_name")
          end
        end

        context "when started_on is after the earliest compensation" do
          let(:employee) { create(:employee, started_on: Date.new(2024, 1, 1)) }
          let(:id) { employee.id }
          let(:body) { { started_on: "2024-06-01" } }
          before { create(:compensation_record, employee: employee, effective_date: Date.new(2024, 1, 1)) }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("started_on")
          end
        end

        context "when the country is not ISO-2" do
          let(:id) { create(:employee).id }
          let(:body) { { country: "USA" } }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("country")
          end
        end
      end
    end
  end

  path "/api/v1/employees/import" do
    post "Import a directory CSV" do
      tags "Employees"
      consumes "multipart/form-data"
      produces "application/json"
      description "Creates employees and effective-dated compensation. Existing emails are updated. Max 5 MB."
      parameter name: :file, in: :formData, schema: { type: :string, format: :binary }, required: true

      response "200", "imported" do
        schema "$ref" => "#/components/schemas/ImportResponse"
        let(:file) { Rack::Test::UploadedFile.new(Rails.root.join("spec/fixtures/files/directory.csv"), "text/csv") }
        before { sign_in_hr }

        run_test! do |response|
          expect(api_data).to include("employees" => 2, "compensation_records" => 3, "skipped" => 0)
        end
      end

      response "401", "not signed in" do
        schema "$ref" => "#/components/schemas/Error"
        let(:file) { Rack::Test::UploadedFile.new(Rails.root.join("spec/fixtures/files/directory.csv"), "text/csv") }
        run_test!
      end

      response "422", "cannot import" do
        schema "$ref" => "#/components/schemas/Error"
        before { sign_in_hr }

        context "when the file is missing" do
          let(:file) { nil }

          run_test! do |response|
            expect(api_error).to include("message" => "file is required")
          end
        end

        context "when the file is not a CSV" do
          let(:file) do
            path = Rails.root.join("tmp/swagger_import.txt")
            File.write(path, "nope")
            Rack::Test::UploadedFile.new(path, "text/plain")
          end

          run_test! do |response|
            expect(api_error).to include("message" => "upload a CSV file")
          end
        end

        context "when the spreadsheet is missing columns" do
          let(:file) do
            path = Rails.root.join("tmp/swagger_import_bad.csv")
            File.write(path, "first_name,email\nAda,ada@acme.test\n")
            Rack::Test::UploadedFile.new(path, "text/csv")
          end

          run_test! do |response|
            expect(api_error.fetch("message")).to include("missing columns")
          end
        end
      end
    end
  end

  path "/api/v1/employees/{id}/offboard" do
    parameter name: :id, in: :path, schema: { type: :string, format: :uuid }

    patch "Mark a leaver" do
      tags "Employees"
      consumes "application/json"
      produces "application/json"
      description "Sets status to left. left_on must be ISO-8601 and cover every compensation date."
      parameter name: :body, in: :body, schema: { "$ref" => "#/components/schemas/OffboardRequest" }

      response "200", "marked left" do
        schema "$ref" => "#/components/schemas/OffboardResponse"
        let(:employee) { create(:employee) }
        let(:id) { employee.id }
        let(:body) { { left_on: "2025-06-01" } }
        before do
          sign_in_hr
          create(:compensation_record, employee: employee)
        end

        run_test! do |response|
          expect(api_data.fetch("employee")).to include("status" => "left", "left_on" => "2025-06-01")
        end
      end

      response "401", "not signed in" do
        schema "$ref" => "#/components/schemas/Error"
        let(:id) { create(:employee).id }
        let(:body) { { left_on: "2025-06-01" } }
        run_test!
      end

      response "404", "unknown employee" do
        schema "$ref" => "#/components/schemas/Error"
        let(:id) { SecureRandom.uuid }
        let(:body) { { left_on: "2025-06-01" } }
        before { sign_in_hr }

        run_test!
      end

      response "422", "cannot offboard" do
        schema "$ref" => "#/components/schemas/ValidationError"
        before { sign_in_hr }

        context "when the employee already left" do
          let(:id) { create(:employee, :left).id }
          let(:body) { { left_on: "2025-07-01" } }

          run_test! do |response|
            expect(api_error).to include("message" => "already left")
          end
        end

        context "when left_on is blank" do
          let(:id) { create(:employee).id }
          let(:body) { { left_on: "" } }

          run_test! do |response|
            expect(api_error).to include("message" => "left_on is required")
          end
        end

        context "when left_on is not ISO-8601" do
          let(:id) { create(:employee).id }
          let(:body) { { left_on: "June 1" } }

          run_test! do |response|
            expect(api_error).to include("message" => "left_on is invalid")
          end
        end

        context "when left_on is before the latest compensation" do
          let(:employee) { create(:employee) }
          let(:id) { employee.id }
          let(:body) { { left_on: "2025-03-01" } }
          before { create(:compensation_record, employee: employee, effective_date: Date.new(2025, 4, 1)) }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("left_on")
          end
        end

        context "when left_on is before started_on" do
          let(:id) { create(:employee, started_on: Date.new(2024, 1, 1)).id }
          let(:body) { { left_on: "2023-12-01" } }

          run_test! do |response|
            expect(api_error.fetch("details")).to include("left_on")
          end
        end
      end
    end
  end
end
